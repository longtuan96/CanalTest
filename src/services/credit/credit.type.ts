import { z } from "zod";

export const CreditCastSchema = z.union([
  z.object({
    adult: z.boolean(),
    gender: z.number(),
    id: z.number(),
    known_for_department: z.string(),
    name: z.string(),
    original_name: z.string(),
    popularity: z.number(),
    profile_path: z.string(),
    cast_id: z.number(),
    character: z.string(),
    credit_id: z.string(),
    order: z.number(),
  }),
  z.object({
    adult: z.boolean(),
    gender: z.number(),
    id: z.number(),
    known_for_department: z.string(),
    name: z.string(),
    original_name: z.string(),
    popularity: z.number(),
    profile_path: z.null(),
    cast_id: z.number(),
    character: z.string(),
    credit_id: z.string(),
    order: z.number(),
  }),
]);
export type CreditCast = z.infer<typeof CreditCastSchema>;

export const CreditCrewSchema = z.union([
  z.object({
    adult: z.boolean(),
    gender: z.number(),
    id: z.number(),
    known_for_department: z.string(),
    name: z.string(),
    original_name: z.string(),
    popularity: z.number(),
    profile_path: z.string(),
    credit_id: z.string(),
    department: z.string(),
    job: z.string(),
  }),
  z.object({
    adult: z.boolean(),
    gender: z.number(),
    id: z.number(),
    known_for_department: z.string(),
    name: z.string(),
    original_name: z.string(),
    popularity: z.number(),
    profile_path: z.null(),
    credit_id: z.string(),
    department: z.string(),
    job: z.string(),
  }),
]);
export type CreditCrew = z.infer<typeof CreditCrewSchema>;

export const GetCreditsResponseSchema = z.object({
  id: z.number(),
  cast: z.array(CreditCastSchema),
  crew: z.array(CreditCrewSchema),
});

export type GetCreditResponse = z.infer<typeof GetCreditsResponseSchema>;

export const GetCreditRequestSchema = z.object({
  movieId: z.string().optional(),
});
export type GetCreditRequest = z.infer<typeof GetCreditRequestSchema>;
