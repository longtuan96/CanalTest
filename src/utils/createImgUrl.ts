import MoviePlaceHolder from "../assets/landscape-placeholder-svgrepo-com.svg";
import PeoplePlaceHolder from "../assets/person-svgrepo-com.svg";

export default function createImgUrl(
  type: "people" | "movie",
  path?: string | null,
  size = "original",
) {
  if (!path) return type === "people" ? PeoplePlaceHolder : MoviePlaceHolder;
  return `${import.meta.env.VITE_THEMOVIEDB_IMG_URL}/${size}${path}`;
}
