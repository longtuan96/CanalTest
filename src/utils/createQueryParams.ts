export default (
  queryParams: Record<string, string | number | null | undefined> = {},
) => ({
  ...queryParams,
  api_key: import.meta.env.VITE_THEMOVIEDB_KEY,
});
