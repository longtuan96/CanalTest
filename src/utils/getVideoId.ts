import { VideoItem } from "../services/video/video.type";

export default function getVideoId(data: VideoItem[] | undefined) {
  if (data) {
    return data.find((item) => item.name === "Official Trailer")?.key;
  } else return undefined;
}
