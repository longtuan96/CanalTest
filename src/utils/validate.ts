import { ZodObject, ZodRawShape } from "zod";

export default function <T>(schema: ZodObject<ZodRawShape>, data: T) {
  const parseResult = schema.safeParse(data);
  if (!parseResult.success) {
    throw Error("Validation Failed!");
  }
  return parseResult.data as T;
}
