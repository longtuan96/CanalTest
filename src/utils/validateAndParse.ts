import { ZodObject, ZodRawShape } from "zod";

export default function validateAndParse<T>(
  schema: ZodObject<ZodRawShape>,
  data: T,
) {
  const parseResult = schema.safeParse(data);
  if (!parseResult.success) {
    throw Error("Validation Failed!");
  }
  return parseResult.data as T;
}
