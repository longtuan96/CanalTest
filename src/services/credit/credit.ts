import { useQuery } from "@tanstack/react-query";
import validateAndParse from "../../utils/validateAndParse";
import { api } from "../api";
import {
  GetCreditRequest,
  GetCreditResponse,
  GetCreditsResponseSchema,
} from "./credit.type";
import { QUERY_KEYS } from "../../constants/keys";
import createQueryParams from "../../utils/createQueryParams";

const { CREDITS } = QUERY_KEYS;
const getCredit = async (payload: GetCreditRequest) => {
  const response = await api.get<GetCreditResponse>(
    `/movie/${payload.movieId}/credits`,
    { params: createQueryParams() },
  );
  return validateAndParse(GetCreditsResponseSchema, response.data);
};

export const useGetCredit = (payload: GetCreditRequest) =>
  useQuery({
    queryKey: [CREDITS, payload],
    queryFn: () => getCredit(payload),
  });
