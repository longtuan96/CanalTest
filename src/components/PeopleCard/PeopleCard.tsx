import { CreditCast } from "../../services/credit/credit.type";
import createImgUrl from "../../utils/createImgUrl";

type PeopleCardProp = {
  data: CreditCast;
};

export const PeopleCard = ({ data }: PeopleCardProp) => {
  return (
    <div className=" min-w-40  rounded-lg bg-white">
      <div id="card-image" className=" min-h-60">
        <img
          className="h-full w-full rounded-t-lg object-fill"
          alt={data.name}
          src={createImgUrl("people", data.profile_path)}
        />
      </div>
      <div id="card-content" className="flex flex-col rounded-b-md text-black">
        <strong className="px-2">{data.name}</strong>
        <span className="px-2 text-sm">{data.character}</span>
      </div>
    </div>
  );
};
