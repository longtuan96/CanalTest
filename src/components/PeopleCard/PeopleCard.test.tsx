import { describe, expect, test } from "vitest";
import { getTree, render, screen } from "../../test/utils";
import { PeopleCard } from "./PeopleCard";
import { creditMockData } from "../../services/credit/credit.mock";

describe("People Card", () => {
  test("render correctly", () => {
    expect(
      getTree(<PeopleCard data={creditMockData.cast[0]} />),
    ).toMatchSnapshot();
  });
  test("alt text must be set", () => {
    render(<PeopleCard data={creditMockData.cast[0]} />);
    const result = screen.getByAltText(creditMockData.cast[0].name);
    expect(result).toBeTruthy();
  });
});
