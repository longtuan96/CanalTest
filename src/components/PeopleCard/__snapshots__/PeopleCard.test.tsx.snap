// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`People Card > render correctly 1`] = `
<div
  className="rounded-lg bg-white"
  style={
    {
      "minWidth": "20%",
    }
  }
>
  <div
    className="min-h-fit w-full"
    id="card-image"
  >
    <img
      alt="Rebecca Hall"
      className="h-full w-full rounded-t-lg object-fill"
      src="https://image.tmdb.org/t/p//original/coC58ANiDbqRIyle5zEl9QDektf.jpg"
    />
  </div>
  <div
    className="flex flex-col rounded-b-md text-black"
    id="card-content"
  >
    <strong
      className="px-2"
    >
      Rebecca Hall
    </strong>
    <span
      className="px-2 text-sm"
    >
      Dr. Ilene Andrews
    </span>
  </div>
</div>
`;
